File: frontend/src/lib/reportPdf.js
// src/lib/reportPdf.js
import { jsPDF } from "jspdf";
import autoTable from "jspdf-autotable";
import { formatCurrency } from "@/lib/formatters";

const num = (v) => Number(v || 0);

export function downloadIncomeStatementPdf(data, { period = "" } = {}) {
  const d = data || {};
  const revenue = num(d.total_revenue ?? d.revenue);
  const cogs = num(d.cost_of_goods_sold ?? d.costOfGoodsSold);
  const grossProfit =
    d.gross_profit != null ? num(d.gross_profit) : revenue - cogs;
  const opex = num(d.operating_expenses ?? d.operatingExpenses);
  const netProfit =
    d.net_profit != null ? num(d.net_profit) : grossProfit - opex;
  const margin =
    d.profit_margin_pct != null
      ? num(d.profit_margin_pct)
      : revenue
      ? (netProfit / revenue) * 100
      : 0;

  const doc = new jsPDF({ unit: "pt", format: "a4" });
  const pageWidth = doc.internal.pageSize.getWidth();

  // Title
  doc.setFontSize(18);
  doc.setFont("helvetica", "bold");
  doc.text("Income & Expense Statement", pageWidth / 2, 50, {
    align: "center",
  });

  // Subtitle (date / period)
  doc.setFontSize(10);
  doc.setFont("helvetica", "normal");
  doc.setTextColor(120);
  const dateStr = new Date().toLocaleDateString();
  doc.text(
    `Generated: ${dateStr}${period ? "   |   Period: " + period : ""}`,
    pageWidth / 2,
    68,
    { align: "center" }
  );
  doc.setTextColor(0);

  const bold = { fontStyle: "bold" };

  autoTable(doc, {
    startY: 92,
    head: [["Description", "Amount"]],
    body: [
      ["Total Revenue / Sales", formatCurrency(revenue)],
      ["Cost of Goods Sold (Purchases)", `(${formatCurrency(cogs)})`],
      [
        { content: "Gross Profit", styles: bold },
        { content: formatCurrency(grossProfit), styles: bold },
      ],
      ["Operating Expenses", `(${formatCurrency(opex)})`],
      [
        { content: netProfit >= 0 ? "Net Income / Profit" : "Net Loss", styles: bold },
        { content: formatCurrency(netProfit), styles: bold },
      ],
      [
        { content: "Profit Margin", styles: bold },
        { content: `${margin.toFixed(2)}%`, styles: bold },
      ],
    ],
    columnStyles: { 1: { halign: "right" } },
    headStyles: { fillColor: [17, 153, 142] }, // teal
    styles: { fontSize: 11, cellPadding: 8 },
    theme: "grid",
  });

  doc.save(`income-statement-${Date.now()}.pdf`);
}

File: mobile app/lib/core/report_pdf.dart
// lib/core/report_pdf.dart
import 'package:pdf/pdf.dart';
import 'package:pdf/widgets.dart' as pw;
import 'package:printing/printing.dart';

String _lkr(num v) => "LKR ${v.toStringAsFixed(2)}";

/// Builds the income statement PDF and opens the share / save / print sheet.
Future<void> shareIncomeStatementPdf(Map<String, dynamic> data) async {
  final num rev = data["total_revenue"] ?? 0;
  final num cogs = data["cost_of_goods_sold"] ?? 0;
  final num gross = data["gross_profit"] ?? (rev - cogs);
  final num opex = data["operating_expenses"] ?? 0;
  final num net = data["net_profit"] ?? (gross - opex);
  final num margin =
      data["profit_margin_pct"] ?? (rev != 0 ? (net / rev) * 100 : 0);

  final doc = pw.Document();

  doc.addPage(
    pw.Page(
      pageFormat: PdfPageFormat.a4,
      margin: const pw.EdgeInsets.all(32),
      build: (context) {
        return pw.Column(
          crossAxisAlignment: pw.CrossAxisAlignment.stretch,
          children: [
            pw.Center(
              child: pw.Text(
                "Income & Expense Statement",
                style: pw.TextStyle(
                  fontSize: 20,
                  fontWeight: pw.FontWeight.bold,
                ),
              ),
            ),
            pw.SizedBox(height: 4),
            pw.Center(
              child: pw.Text(
                "Generated: ${DateTime.now().toString().split('.').first}",
                style: const pw.TextStyle(
                  fontSize: 10,
                  color: PdfColors.grey700,
                ),
              ),
            ),
            pw.SizedBox(height: 24),
            pw.Table(
              border: pw.TableBorder.all(color: PdfColors.grey400, width: 0.5),
              columnWidths: {
                0: const pw.FlexColumnWidth(3),
                1: const pw.FlexColumnWidth(2),
              },
              children: [
                _row("Description", "Amount", header: true),
                _row("Total Revenue / Sales", _lkr(rev)),
                _row("Cost of Goods Sold (Purchases)", "(${_lkr(cogs)})"),
                _row("Gross Profit", _lkr(gross), bold: true),
                _row("Operating Expenses", "(${_lkr(opex)})"),
                _row(
                  net >= 0 ? "Net Income / Profit" : "Net Loss",
                  _lkr(net),
                  bold: true,
                ),
                _row("Profit Margin", "${margin.toStringAsFixed(2)}%",
                    bold: true),
              ],
            ),
          ],
        );
      },
    ),
  );

  await Printing.sharePdf(
    bytes: await doc.save(),
    filename: "income-statement.pdf",
  );
}

pw.TableRow _row(String a, String b, {bool header = false, bool bold = false}) {
  final style = pw.TextStyle(
    fontWeight: (header || bold) ? pw.FontWeight.bold : pw.FontWeight.normal,
    fontSize: header ? 12 : 11,
    color: header ? PdfColors.white : PdfColors.black,
  );
  return pw.TableRow(
    decoration: header
        ? const pw.BoxDecoration(color: PdfColor.fromInt(0xFF11998E)) // teal
        : null,
    children: [
      pw.Padding(
        padding: const pw.EdgeInsets.all(8),
        child: pw.Text(a, style: style),
      ),
      pw.Padding(
        padding: const pw.EdgeInsets.all(8),
        child: pw.Text(b, style: style, textAlign: pw.TextAlign.right),
      ),
    ],
  );
}